src/snowflake/cli/api/metrics.py

```diff
@@ -54,18 +70,96 @@ class CLICounterField:
     PACKAGE_SCRIPTS = f"{_TypePrefix.FEATURES}.{_DomainPrefix.APP}.package_scripts"
 
 
+@dataclass
+class CLIMetricsSpan:
+    """
+    class for holding metrics span data and encapsulating related operations
+    """
+
+    # keys for dict representation
+    ID_KEY: ClassVar[str] = "id"
+    NAME_KEY: ClassVar[str] = "name"
+    PARENT_KEY: ClassVar[str] = "parent"
+    PARENT_ID_KEY: ClassVar[str] = "parent_id"
+    START_TIME_KEY: ClassVar[str] = "start_time"
+    EXECUTION_TIME_KEY: ClassVar[str] = "execution_time"
+    ERROR_KEY: ClassVar[str] = "error"
+
+    # ensure we get unique ids for each step for the parent-child link in case of steps with the same name
+    step_id: str = field(init=False, default_factory=lambda: uuid.uuid4().hex)
+    name: str
+    # start time of this step relative to when the command first started executing
+    start_time: float
+    parent: Optional[CLIMetricsSpan] = field(default=None)
+    execution_time: Optional[float] = field(init=False, default=None)
+    error: Optional[BaseException] = field(init=False, default=None)
+
+    # start time of the step from the monotonic clock in order to calculate execution time
+    _monotonic_start: float = field(init=False, default_factory=time.monotonic)
+
+    def __post_init__(self):
+        if not self.name:
+            raise CLIMetricsInvalidUsageError("step name must not be empty")
+
+    def finish(self, error: Optional[BaseException] = None) -> None:
+        """
+        Sets the execution time and (optionally) error raised for the span
+
+        If already called, this method is a no-op
+        """
+        if self.execution_time is not None:
+            return
+
+        if error:
+            self.error = error
+
+        self.execution_time = time.monotonic() - self._monotonic_start
+
+    def to_dict(self) -> Dict:
+        """
+        Custom dict conversion function to be used for reporting telemetry, with only the required fields
+        """
+
+        return {
+            self.ID_KEY: self.step_id,
+            self.NAME_KEY: self.name,
+            self.START_TIME_KEY: self.start_time,
+            self.PARENT_KEY: self.parent.name if self.parent is not None else None,
+            self.PARENT_ID_KEY: self.parent.step_id
+            if self.parent is not None
+            else None,
+            self.EXECUTION_TIME_KEY: self.execution_time,
+            self.ERROR_KEY: type(self.error).__name__ if self.error else None,
+        }
+
+
+@dataclass
```

Comment:
Converted to dataclass so that we get equality comparison and cloning for free to help with forking the cli context